pyteal/ast/abi/util_test.py

```diff
@@ -772,3 +772,250 @@ def test_sdk_type_specs_from_signature(sig_str, sig_args, sig_rets):
     args, ret = abi.type_specs_from_signature(sig_str)
     assert args == sig_args
     assert ret == sig_rets
+
+
+class NamedTDecl(abi.NamedTuple):
+    a: abi.Field[abi.Uint64]
+    b: abi.Field[
+        abi.Tuple3[abi.PaymentTransaction, abi.Address, abi.StaticBytes[Literal[16]]]
+    ]
+    c: abi.Field[abi.Transaction]
+
+
+class NamedTComp0(abi.NamedTuple):
+    a0: abi.Field[abi.String]
+    a1: abi.Field[abi.Address]
+
+
+class NamedTComp1(abi.NamedTuple):
+    b0: abi.Field[abi.DynamicBytes]
+    b1: abi.Field[abi.StaticBytes[Literal[32]]]
+
+
+class NamedTComp2(abi.NamedTuple):
+    b1: abi.Field[abi.Address]
+    b0: abi.Field[abi.DynamicBytes]
+
+
+class SafeBidirectional(NamedTuple):
+    xs: list[abi.TypeSpec]
+
+
+SAFE_BIDIRECTIONAL_TEST_CASES: list[SafeBidirectional] = (
+    [
+        SafeBidirectional(
+            [
+                abi.type_spec_from_annotation(abi.StaticArray[abi.Byte, Literal[10]]),
+                abi.type_spec_from_annotation(abi.StaticBytes[Literal[10]]),
+            ],
+        ),
+        SafeBidirectional([abi.BoolTypeSpec(), abi.BoolTypeSpec()]),
+        SafeBidirectional([abi.StringTypeSpec(), abi.StringTypeSpec()]),
+        SafeBidirectional([abi.AddressTypeSpec(), abi.AddressTypeSpec()]),
+        SafeBidirectional(
+            [
+                abi.type_spec_from_annotation(abi.DynamicBytes),
+                abi.type_spec_from_annotation(abi.DynamicArray[abi.Byte]),
+            ]
+        ),
+        SafeBidirectional(
+            [
+                abi.type_spec_from_annotation(
+                    abi.Tuple3[
+                        abi.Uint64,
+                        abi.Tuple3[
+                            abi.PaymentTransaction,
+                            abi.Address,
+                            abi.StaticArray[abi.Byte, Literal[16]],
+                        ],
+                        abi.Transaction,
+                    ]
+                ),
+                abi.type_spec_from_annotation(NamedTDecl),
+                abi.type_spec_from_annotation(NamedTDecl),
+            ]
+        ),
+    ]
+    + [
+        SafeBidirectional([spec, spec])
+        for spec in abi.ReferenceTypeSpecs + abi.TransactionTypeSpecs
+    ]
+    + [
+        SafeBidirectional([spec_t(), spec_t()])
```

Comment:
@ahangsu Request for clarification:  There's a few `SafeBidirectional` test cases that duplicate the value under test.  Examples include the highlighted line + these examples (from up above):
```
        SafeBidirectional([abi.BoolTypeSpec(), abi.BoolTypeSpec()]),
        SafeBidirectional([abi.StringTypeSpec(), abi.StringTypeSpec()]),
        SafeBidirectional([abi.AddressTypeSpec(), abi.AddressTypeSpec()]),
```

Can you help me understand how duplicating the value rather than showing a single value enriches the test?

If there's no reason to support value duplication, then I'd optionally consider changing from list to set.

Reply:
moving to set is making sense, it was mistaken to have duplicate cases.

There are some difficulty in moving to set, for some type spec instances are not hashable. Need to figure out how to get it work.

Reply:
@ahangsu While it'd be nice to understand the behavior, I think https://github.com/algorand/pyteal/pull/540/commits/4315dd0d90da2c8a1e428b551ce7c0e160faf8a2 is sufficient for now.  Thanks for addressing the concern - Resolving.